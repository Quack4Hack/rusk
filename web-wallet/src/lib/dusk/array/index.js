export { default as arraysEqual } from "./arraysEqual";
export { default as shuffleArray } from "./shuffleArray";
export { default as findFirstNMatches } from "./findFirstNMatches";
