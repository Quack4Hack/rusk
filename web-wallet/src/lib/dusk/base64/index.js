export { default as base64ToUTF8 } from "./base64ToUTF8";
export { default as base64ToBytes } from "./base64ToBytes";
export { default as bytesToBase64 } from "./bytesToBase64";
export { default as utf8ToBase64 } from "./utf8ToBase64";
