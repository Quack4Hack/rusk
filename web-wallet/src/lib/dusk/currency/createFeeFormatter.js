/**
 * Creates a locale aware currency formatter for fees
 *
 * @param {String} locale A BCP 47 language tag
 * @returns {(value: number | bigint) => string}
 */
const createFormatter = (locale) => {
	const formatter = new Intl.NumberFormat(locale, {
		maximumFractionDigits: 9,
		minimumFractionDigits: 2
	});

	return value => formatter.format(value);
};

export default createFormatter;
