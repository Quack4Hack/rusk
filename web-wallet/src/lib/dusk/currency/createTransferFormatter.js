/**
 * Creates a locale aware currency formatter for fiat or DUSK
 *
 * @param {String} locale A BCP 47 language tag
 * @returns {(value: number | bigint) => string}
 */
const createFormatter = (locale) => {
  const formatter = new Intl.NumberFormat(locale, {
    maximumFractionDigits: 9,
    minimumFractionDigits: 2,
    signDisplay: "exceptZero",
  });

  return (value) => formatter.format(value);
};

export default createFormatter;
