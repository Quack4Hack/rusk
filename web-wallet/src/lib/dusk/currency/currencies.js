const currencies = [
  { code: "AED", currency: "United Arab Emirates Dirham" },
  { code: "ARS", currency: "Argentine Peso" },
  { code: "AUD", currency: "Australian Dollar" },
  { code: "BDT", currency: "Bangladeshi Taka" },
  { code: "BHD", currency: "Bahraini Dinar" },
  { code: "BMD", currency: "Bermudian Dollar" },
  { code: "BRL", currency: "Brazilian Real" },
  { code: "CAD", currency: "Canadian Dollar" },
  { code: "CFH", currency: "Swiss Franc" },
  { code: "CLP", currency: "Chilean Peso" },
  { code: "CNY", currency: "Renminbi" },
  { code: "CZK", currency: "Czech Koruna" },
  { code: "DKK", currency: "Danish Krone" },
  { code: "EUR", currency: "Euro" },
  { code: "GBP", currency: "Pound Sterling" },
  { code: "HKD", currency: "Hong Kong Dollar" },
  { code: "HUF", currency: "Hungarian Forint" },
  { code: "IDR", currency: "Indonesian Rupiah" },
  { code: "ILS", currency: "Israeli New Shekel" },
  { code: "INR", currency: "Indian Rupee" },
  { code: "JPY", currency: "Japanese Yen" },
  { code: "KRW", currency: "South Korean Won" },
  { code: "KWD", currency: "Kuwaiti Dinar" },
  { code: "LKR", currency: "Sri Lankan Rupee" },
  { code: "MMK", currency: "Myanmar Kyat" },
  { code: "MXN", currency: "Mexican Peso" },
  { code: "MYR", currency: "Malaysian Ringgit" },
  { code: "NGN", currency: "Nigerian Naira" },
  { code: "NOK", currency: "Norwegian Krone" },
  { code: "NZD", currency: "New Zealand Dollar" },
  { code: "PHP", currency: "Philippine Peso" },
  { code: "PKR", currency: "Pakistani Rupee" },
  { code: "PLN", currency: "Polish Złoty" },
  { code: "RUB", currency: "Russian Ruble" },
  { code: "SAR", currency: "Saudi Riyal" },
  { code: "SEK", currency: "Swedish Krona" },
  { code: "SGD", currency: "Singapore Dollar" },
  { code: "THB", currency: "Thai Baht" },
  { code: "TRY", currency: "Turkish Lira" },
  { code: "TWD", currency: "New Taiwan Dollar" },
  { code: "UAH", currency: "Ukrainian Hryvnia" },
  { code: "USD", currency: "United States Dollar" },
  { code: "VEF", currency: "Venezuelan Bolívar Fuerte" },
  { code: "VND", currency: "Vietnamese Dồng" },
  { code: "ZAR", currency: "South African Rand" },
];

export default currencies;
