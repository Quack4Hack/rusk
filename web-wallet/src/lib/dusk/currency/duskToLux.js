/**
 * Temporary conversion function until
 * `dusk-wallet-js` exposes its own.
 *
 * @param {number} n
 * @returns {number}
 */
const duskToLux = n => n * 1e9;

export default duskToLux;
