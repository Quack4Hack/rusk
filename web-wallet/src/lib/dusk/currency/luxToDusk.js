const scaleFactorAsNumber = 1e9;
const scaleFactorAsBigInt = BigInt(scaleFactorAsNumber);

/**
 * Temporary conversion: in the near future
 * we will accept only BigInt as input.
 *
 * @param {bigint} n
 * @returns {number}
 */
function luxToDusk(n) {
  const integerPart = Number(n / scaleFactorAsBigInt);
  const decimalPart = Number(n % scaleFactorAsBigInt);

  return integerPart + decimalPart / scaleFactorAsNumber;
}

export default luxToDusk;
