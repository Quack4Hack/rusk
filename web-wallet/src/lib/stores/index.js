export { default as operationsStore } from "./operationsStore";
export { default as settingsStore } from "./settingsStore";
export { default as walletStore } from "./walletStore";
