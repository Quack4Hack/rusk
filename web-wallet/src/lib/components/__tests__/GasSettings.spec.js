import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";

import { GasSettings } from "..";
import { get } from "svelte/store";
import { settingsStore } from "$lib/stores";
import { createCurrencyFormatter } from "$lib/dusk/currency";

describe("GasSettings", () => {
  const settings = get(settingsStore);
  const duskFormatter = createCurrencyFormatter(settings.language, "DUSK", 9);
  const fee = settings.gasPrice * settings.gasLimit;

  const baseProps = {
    fee: fee,
    formatter: duskFormatter,
    limit: 20000000n,
    limitLower: 10000000n,
    limitUpper: 1000000000n,
    price: 1n,
    priceLower: 1n,
  };

  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  it("renders the GasSettings component closed", () => {
    const { container } = render(GasSettings, baseOptions);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the GasSettings component opened", async () => {
    const { container, getByRole } = render(GasSettings, baseOptions);

    const next = getByRole("button", { name: "EDIT" });

    await fireEvent.click(next);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('checks "gasSettings" event is dispatched on click with the correct event data', async () => {
    const eventHandler = vi.fn();
    const { component, getByRole, getAllByRole } = render(
      GasSettings,
      baseOptions
    );
    const editButton = getByRole("button", { name: "EDIT" });

    expect(() => getAllByRole("textbox")).toThrow();

    await fireEvent.click(editButton);

    component.$on("gasSettings", eventHandler);

    const [priceInput, limitInput] = getAllByRole("textbox");

    await fireEvent.input(limitInput, {
      target: { value: baseProps.limitLower },
    });

    expect(eventHandler).toHaveBeenCalledTimes(1);
    expect(eventHandler.mock.lastCall?.[0].detail).toStrictEqual({
      limit: baseProps.limitLower,
      price: baseProps.price,
    });

    await fireEvent.input(priceInput, {
      target: { value: baseProps.price * 2n },
    });

    expect(eventHandler).toHaveBeenCalledTimes(2);
    expect(eventHandler.mock.lastCall?.[0].detail).toStrictEqual({
      limit: baseProps.limitLower,
      price: baseProps.price * 2n,
    });
  });
});
