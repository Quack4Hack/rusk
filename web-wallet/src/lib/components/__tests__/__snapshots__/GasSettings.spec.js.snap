// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GasSettings > renders the GasSettings component closed 1`] = `
<div
  class="gas-settings svelte-1daay07"
>
  <dl
    class="gas-settings__edit svelte-1daay07"
  >
    <div
      class="gas-fee svelte-1c6b481"
    >
      <span>
        Fee:
      </span>
       
      <div
        class="gas-fee__amount svelte-1c6b481"
      >
        <div
          class="gas-fee__amount-value svelte-1c6b481"
        >
          <span
            class="svelte-1c6b481"
          >
            max
          </span>
           
          <span
            class="svelte-1c6b481"
          >
            0.020000000
          </span>
        </div>
         
        <svg
          class="dusk-icon dusk-icon--size--default dusk-amount__icon"
          data-tooltip-id="main-tooltip"
          data-tooltip-text="DUSK"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
          />
        </svg>
        
      </div>
    </div>
    <dd>
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          EDIT
        </span>
      </button>
    </dd>
  </dl>
   
</div>
`;

exports[`GasSettings > renders the GasSettings component opened 1`] = `
<div
  class="gas-settings svelte-1daay07"
>
  <dl
    class="gas-settings__edit svelte-1daay07"
  >
    <div
      class="gas-fee svelte-1c6b481"
    >
      <span>
        Fee:
      </span>
       
      <div
        class="gas-fee__amount svelte-1c6b481"
      >
        <div
          class="gas-fee__amount-value svelte-1c6b481"
        >
          <span
            class="svelte-1c6b481"
          >
            max
          </span>
           
          <span
            class="svelte-1c6b481"
          >
            0.020000000
          </span>
        </div>
         
        <svg
          class="dusk-icon dusk-icon--size--default dusk-amount__icon"
          data-tooltip-id="main-tooltip"
          data-tooltip-text="DUSK"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
          />
        </svg>
        
      </div>
    </div>
    <dd>
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          CLOSE
        </span>
      </button>
    </dd>
  </dl>
   
  <div
    class="gas-settings svelte-1daay07"
    style="animation: __svelte_276924520_0 400ms linear 0ms 1 both;"
  >
    <label
      class="gas-control svelte-1p4ekyk"
    >
      <span
        class="gas-control__label svelte-1p4ekyk"
      >
        Price: (lux)
      </span>
       
      <input
        class="dusk-textbox dusk-textbox-number gas-control__input"
        max="20000000"
        min="1"
        required=""
        type="number"
      />
      
    </label>
     
    <label
      class="gas-control svelte-1p4ekyk"
    >
      <span
        class="gas-control__label svelte-1p4ekyk"
      >
        Gas Limit: (unit)
      </span>
       
      <input
        class="dusk-textbox dusk-textbox-number gas-control__input"
        max="1000000000"
        min="10000000"
        required=""
        type="number"
      />
      
    </label>
  </div>
</div>
`;
